pkg/codegen/dotnet/gen_program.go

```diff
@@ -157,6 +214,28 @@ func GenerateProgram(program *pcl.Program) (map[string][]byte, hcl.Diagnostics,
 	return GenerateProgramWithOptions(program, defaultOptions)
 }
 
+func collectPackages(program *pcl.Program, seenPackages map[string]*schema.Package) error {
```

Comment:
> To keep existing code working as much as possible and avoid replication, could we fold this functionality into program.PackageSnapshots()?

I wasn't sure if the behaviour would remain the same for program-gen in other languages that use the feature so I only implemented it here in dotnet

> Also, why program.PackageSnapshots() instead of program.PackageReferences()? Having incomplete schema.Packages floating around seems more error prone then converting them into schema.PackageReferences.

I am not sure tbh, only extended the behaviour to traverse child programs within components rather than the top-level program alone.